src/appleseed/renderer/modeling/phasefunction/henyeyphasefunction.cpp

```diff
@@ -0,0 +1,323 @@
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float                       distance,
+        foundation::Vector3f&       incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const foundation::Vector3f outgoing =
+            foundation::Vector3f(foundation::normalize(volume_ray.m_dir));
+        const foundation::Basis3f basis(outgoing);
+
+        //
+        // x = 1/2g * (1 + g^2 - [(1 - g^2) / (1 + g*s)]^2),
+        // where x is cos(phi) and s is uniform random sample from [-1; 1].
+        //
+
+        const float g = values->m_average_cosine;
+        const float sqr_g = g * g;
+
+        sampling_context.split_in_place(1, 2);
+        const float s = 2.0f * sampling_context.next2<float>() - 1.0f;
+
+        const float t = (1.0f - sqr_g) / (1.0f + g*s);
+        const float cosine = 0.5f * g * (1 + sqr_g - t * t);
+        const float sine = sqrt(1.0f - cosine * cosine);
+
+        const foundation::Vector2f tangent =
+            foundation::sample_circle_uniform(sampling_context.next2<float>());
+
+        incoming =
+            basis.get_tangent_u() * tangent.x * sine +
+            basis.get_tangent_v() * tangent.y * sine +
+            basis.get_normal()    * cosine;
+
+        // Evaluate PDF.
+
+        const float numerator = (1.0f - sqr_g);
+        const float denumerator = powf(1 + sqr_g - 2*g*cosine, -1.5f);
+
+        return foundation::RcpFourPi<float>() * numerator * denumerator;
+    }
+
+    virtual float evaluate(
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float                       distance,
+        const foundation::Vector3f& incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const foundation::Vector3f outgoing =
+            foundation::Vector3f(foundation::normalize(volume_ray.m_dir));
+        const float cosine = foundation::dot(incoming, outgoing);
+
+        //
+        // p(x) = 1/4pi * (1 - g^2) / (1 + g^2 - 2gx)^(3/2),
+        // where x is cos(phi) and g is the average cosine parameter.
+        //
+
+        const float g = values->m_average_cosine;
+        const float sqr_g = g * g;
+
+        const float numerator = (1.0f - sqr_g);
+        const float denumerator = powf(1.0f + sqr_g - 2*g*cosine, -1.5f);
+
+        return foundation::RcpFourPi<float>() * numerator * denumerator;
+    }
+
+    virtual void evaluate_transmission(
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float                       distance,
+        Spectrum&                   spectrum) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+
+        spectrum = foundation::exp(-distance * values->m_extinction_multiplier * values->m_extinction);
```

Comment:
I really want to find a component-wise exponent of a Spectrum here. It is the only way to compute transmission.

Maybe instead of these coefficients we would like to add some intuitive parameters, like we do here: [glassbsdf.cpp#L459](https://github.com/appleseedhq/appleseed/blob/master/src/appleseed/renderer/modeling/bsdf/glassbsdf.cpp#L459)
```c++
    // [2] Volumetric absorption reparameterization, page 5.
    absorption.resize(values->m_volume_transmittance.size());
    const float d = distance / values->m_volume_transmittance_distance;
    for (size_t i = 0, e = absorption.size(); i < e; ++i)
    {
        const float a = log(max(values->m_volume_transmittance[i], 0.01f));
        absorption[i] = exp(a * d);
    }
```